### PR TITLE
Test for  task/:username route

### PR DESCRIPTION
**Pull request by bhk31**

### CODE REVIEW

File: test/integration/tasks.test.js [hunk 1]
```diff
@@ -249,4 +249,33 @@ describe('Tasks', function () {
...
+      chai
+        .request(app)
+        .get('/tasks/sagar')
+        .end((err, res) => {
+          if (err) { return done(err) }
+          expect(res).to.have.status(200)
+          expect(res.body).to.be.a('object')
+          expect(res.body.message).to.equal('Tasks returned successfully!')
+          expect(res.body.tasks).to.be.a('array')
+          return done()
```
Comment (swarajpure) [inline test/integration/tasks.test.js:264]: Can we have stricter assertions on `res.body.tasks`?
What I mean is `expect(res.body.tasks).to.have.all.keys(...taskObjectKeys)`
Where `taskObjectKeys` is an array of the keys of the task object
You can refer this: https://github.com/Real-Dev-Squad/website-backend/blob/09c53d861f1133e9ab2bea3a1421bcec855f8716/test/integration/auction.test.js#L51

Comment (bhk31) [reply test/integration/tasks.test.js:264]: @swarajpure I tried to add assertions for `res.body.tasks` but when I looked into the result of `res.body.tasks`. It outputs some different object which we don't have in any variable stored. This is what it outputs.
`[
  {
    id: '7f4MpynEfkksfRQkcTcG',
    percentCompleted: 10,
    endsOn: '<unix timestamp>',
    isNoteworthy: true,
    dependsOn: [ 'd12', 'd23' ],
    purpose: 'To Test mocha',
    lossRate: { dinero: 1 },
    title: 'Assigned task',
    type: 'Dev | Group',
    completionAward: { dinero: 3, neelam: 300 },
    links: [ 'test1' ],
    assignee: 'user1',
    featureUrl: '<testUrl>',
    startedOn: '<unix timestamp>',
    status: 'active',
    participants: [ 'sagar' ]
  },
  {
    id: 'NSbhS5GCQaI3ePrBVS4k',
    percentCompleted: 10,
    endsOn: 1234,
    isNoteworthy: true,
    lossRate: { dinero: 1 },
    title: 'new-title',
    type: 'feature',
    completionAward: { dinero: 3, neelam: 300 },
    startedOn: 4567,
    status: 'active',
    participants: [ 'sagar' ]
  },
  {
    id: 'eHeuYaOY3iFN3m49l4qg',
    percentCompleted: 10,
    endsOn: 1234,
    isNoteworthy: false,
    dependsOn: [ 'd12', 'd23' ],
    purpose: 'To Test mocha',
    lossRate: { dinero: 1 },
    title: 'Test task',
    type: 'group',
    completionAward: { dinero: 3, neelam: 300 },
    links: [ 'test1' ],
    featureUrl: '<testUrl>',
    startedOn: 54321,
    status: 'completed',
    participants: [ 'sagar' ]
  },
  {
    id: 'H848N7dbBT4l5SLfZf54',
    percentCompleted: 10,
    endsOn: 123,
    lossRate: { dinero: 1 },
    title: 'Test task - Create',
    type: 'feature',
    completionAward: { dinero: 3, neelam: 300 },
    createdBy: 'sagar',
    assignee: 'sagar',
    startedOn: 456,
    participants: [],
    status: 'completed'
  }
]
`

I think its adding elements to tasks dynamically when we execute previous test cases. Could you please help me understand what logic we can apply ?

Comment (swarajpure) [reply test/integration/tasks.test.js:264]: I get it. There are some optional fields.
Can we are least check for assignee field? Making sure it returns the `username` provided.

**Review by swarajpure — CHANGES_REQUESTED** [review]
Can you please also add unit tests for the getUserTasks functions? We need to have robust tests. We've seen tasks API failing a couple of times last month. It's a critical one as it's being used on members site which is seen by recruiters. We don't want our APIs to break in front of them, right? I hope that clears as to why I'm so stubborn about tests 😂

**Comment by bhk31** [comment]
> Can you please also add unit tests for the getUserTasks functions? We need to have robust tests. We've seen tasks API failing a couple of times last month. It's a critical one as it's being used on members site which is seen by recruiters. We don't want our APIs to break in front of them, right? I hope that clears as to why I'm so stubborn about tests 😂

Where should we write test for controller function.

File: test/integration/tasks.test.js (hunk 1 above)
Comment (bhk31) [reply test/integration/tasks.test.js]: Updated the code for this. Please review it.

**Comment by swarajpure** [comment]
> > Can you please also add unit tests for the getUserTasks functions? We need to have robust tests. We've seen tasks API failing a couple of times last month. It's a critical one as it's being used on members site which is seen by recruiters. We don't want our APIs to break in front of them, right? I hope that clears as to why I'm so stubborn about tests joy
> 
> @swarajpure Where should we write test for controller function?

Integration tests are testing middlewares + controllers flow
So we need not write tests for controllers explicitly since they only accept req and res objects

We need to be concerned about model functions since slight variation in function arguments can change the output. Edge cases need to be handled there.

**Comment by swarajpure** [comment]
I'm merging this PR for now.
At least we would have integration tests in place.

New issue created to track unit tests for model functions: #363

**Review by swarajpure — APPROVED** [review]
Merging this for now
Issue #363 for unit tests for model functions

Great job in here man. I know writing tests can be boring sometimes, but when you see some bugs in production, some users unable to see their members profiles/contribution, then we realise how important tests are. Once you realise "why" of tests, they are fun and super super important. 

Don't think of them as boring, you're investing your time in saving potential bugs which could be introduced by developers in future and helping RDS to deliver bug-free experience to our users 🙂

**Comment by bhk31** [comment]
Thanks @swarajpure. It was good learning experience.
